File: panel/src/features/auth/data/repositories/AuthRepositoryImpl.ts
import type { AuthSession } from '../../domain/entities/AuthSession';
import type { AuthUser } from '../../domain/entities/AuthUser';
import { AuthRepository } from '../../domain/repositories/AuthRepository';
import { AuthRemoteDataSource } from '../datasource/AuthRemoteDataSource';
import { AuthSessionModel } from '../models/AuthSessionModel';
import { AuthUserModel } from '../models/AuthUserModel';

export class AuthRepositoryImpl extends AuthRepository {
  constructor(private readonly remote: AuthRemoteDataSource) {
    super();
  }

  async login(input: {
    email: string;
    password: string;
  }): Promise<AuthSession> {
    const data = await this.remote.login(input);
    return AuthSessionModel.fromJson(data);
  }

  async register(input: {
    firstName: string;
    lastName: string;
    email: string;
    password: string;
    role: string;
    phone?: string;
    gymId?: string;
    gymName?: string;
    gymSlug?: string;
  }): Promise<AuthSession> {
    const data = await this.remote.register(input);
    return AuthSessionModel.fromJson(data);
  }

  async me(token: string): Promise<AuthUser> {
    const data = await this.remote.me(token);
    return AuthUserModel.fromJson(data);
  }
}

File: app/lib/features/auth/data/repositories/auth_repository_impl.dart
import '../../domain/entities/auth_session.dart';
import '../../domain/repositories/auth_repository.dart';
import '../datasource/auth_remote_datasource.dart';
import '../models/login_request_model.dart';
import '../models/register_request_model.dart';

class AuthRepositoryImpl implements AuthRepository {
  final AuthRemoteDataSource remoteDataSource;

  AuthRepositoryImpl(this.remoteDataSource);

  @override
  Future<AuthSession> login({
    required String email,
    required String password,
  }) {
    return remoteDataSource.login(
      LoginRequestModel(
        email: email,
        password: password,
      ),
    );
  }

  @override
  Future<AuthSession> register({
    required String firstName,
    required String lastName,
    required String email,
    required String password,
    required String role,
    String? gymId,
    String? gymName,
    String? gymSlug,
    String? phone,
  }) {
    return remoteDataSource.register(
      RegisterRequestModel(
        firstName: firstName,
        lastName: lastName,
        email: email,
        password: password,
        role: role,
        gymId: gymId,
        gymName: gymName,
        gymSlug: gymSlug,
        phone: phone,
      ),
    );
  }
}